en/middleware/micrortps.md

```diff
@@ -0,0 +1,344 @@
+├── microRTPS_transport.cxx
+└── microRTPS_transport.h
+ 0 directories, 4 files
+```
+
+
+## PX4 Firmware: the micro RTPS client
+
+The agent application runs as both a uORB node and as a micro RTPS node. As a uORB node the agent could be subscribed to several topics as well as publish under internal uORB topics. The application receives from a internal publishers the messages, serializes the struct and writes it trough an UART
+device or UDP port selected by the user. Also will be reading from the UART device or UDP port and then publish the info
+to the internal subscribers.
+
+Steps to use the auto generated application:
+
+- Check that the lines **modules/micrortps_bridge/micrortps_client** and **lib/micro-CDR** are present in the *.cmake* config file for the target platform (*cmake/configs/*). This enables the compilation of the client along with the **PX4** firmware:
+
+  ```sh
+  set(config_module_list
+      ...
+      lib/micro-CDR
+      ...
+      # micro RTPS
+      modules/micrortps_bridge/micrortps_client
+      ...
+      )
+  ```
+  > **Note** For Nuttx platforms (e.g. *Pixracer*) the cmake files are **cmake/configs/nuttx_px4fmu-v4_default.cmake**. For the *Snapdragon Flight* platform you can find the cmake configuration here: **cmake/configs/posix_sdflight_default.cmake**.
+  
+- Construct and upload the firmware executing, for example:
+
+  ```sh
+  # For NuttX/Pixhawk flight controllers:
+  $ make px4fmu-v4_default upload
+  ```
+  ```sh
+  # For Snapdragon Flight:
+  $ make eagle_default upload
+  ```
+
+After uploading the firmware, the application can be launched typing its name and passing an variable number of arguments as shown below:
+
+```sh
+> micrortps_client start|stop [options]
+  -t <transport>          [UART|UDP] Default UART
+  -d <device>             UART device. Default /dev/ttyACM0
+  -u <update_time_ms>     Time in ms for uORB subscribed topics update. Default 0
+  -l <loops>              How many iterations will this program have. -1 for infinite. Default 10000
+  -w <sleep_time_ms>      Time in ms for which each iteration sleep. Default 1ms
+  -b <baudrate>           UART device baudrate. Default 460800
+  -p <poll_ms>            Time in ms to poll over UART. Default 1ms
+  -r <reception port>     UDP port for receiving. Default 2019
+  -s <sending port>       UDP port for sending. Default 2020
+```
+
+> **Note** When working with a USB-serial adapter the maximum link speed is used (the `-b` option is ignored).
+  ```sh
+  > micrortps_client start #by default -t UART -d /dev/ttyACM0 -u 0 -l 10000 -w 1 -b 460800 -p 1
+  ```
+
+<span></span>
+> **Note** If the selected UART port is busy, it's possible that the MAVLink application is already being used. If it is the case, you can stop MAVLink from NuttShell by typing:
+  ```sh
+  > mavlink stop-all
+  ```
+ 
+## Fast RTPS: the Micro RTPS agent
+
+The *Fast RTPS* side will be explained taking a *Raspberry Pi* board to run an application as example.
```

Comment:
The text below here to "To create the application, compile the code:" seems to have some duplication and repetition.

What isn't clear is what is the role of the agent, and how it fits with the client. What I guess is that this is the PX4 side code that implements FastRTPS protocol and talks to the protocol on the remote system (e.g. RaPi). I THINK that this is sharing information with the client. So uOrb -> client (serialise) -> agent (rtps on px4)  sends rtps -> port to remote system -> rtps on remote system ....